File: weather.ts
/*
Get Takoma Park's weather coords:
38.9823732,-77.0065528
curl -L https://api.weather.gov/points/38.9823732,-77.0065528

Get Takoma Park's weather forecast:
curl https://api.weather.gov/gridpoints/LWX/97,75/forecast | jq .properties.periods
*/

const FORECAST_URL = "https://api.weather.gov/gridpoints/LWX/97,75/forecast/hourly";
const PUSHOVER_USER = process.env.PUSHOVER_USER;
const PUSHOVER_TOKEN = process.env.PUSHOVER_TOKEN;

// NOAA API Response
interface APIWeatherForecast {
  number: number;
  name: string;
  startTime: string;
  endTime: string;
  isDaytime: boolean;
  temperature: number;
  temperatureUnit: string;
  temperatureTrend: any;
  probabilityOfPrecipitation: ProbabilityOfPrecipitation;
  dewpoint: Dewpoint;
  relativeHumidity: RelativeHumidity;
  windSpeed: string;
  windDirection: string;
  icon: string;
  shortForecast: string;
  detailedForecast: string;
}

interface ProbabilityOfPrecipitation {
  unitCode: string;
  value: number;
}

interface Dewpoint {
  unitCode: string;
  value: number;
}

interface RelativeHumidity {
  unitCode: string;
  value: number;
}

async function getWeather(): Promise<{
  weather: APIWeatherForecast[];
  error: string;
}> {
  try {
    const resp = await fetch(FORECAST_URL);
    const body = await resp.json();
    const periods: APIWeatherForecast[] = body.properties.periods;
    if (resp.status > 299) {
      return {
        weather: [],
        error: `NOAA API returned error - ${resp.status} - ${JSON.stringify(
          resp.body
        )}`,
      };
    }

    return {
      weather: periods,
      error: "",
    };
  } catch (err) {
    return { weather: [], error: `error: ${err}` };
  }
}

async function push(msg: string): Promise<{ error: string }> {
  try {
    const resp = await fetch("https://api.pushover.net/1/messages.json", {
      method: "POST",
      headers: {
        "content-type": "application/json",
      },
      body: JSON.stringify({
        token: PUSHOVER_TOKEN,
        user: PUSHOVER_USER,
        message: msg,
      }),
    });
    if (resp.status > 299) {
      return { error: await resp.text() };
    }
  } catch (err) {
    return { error: `error: ${err}` };
  }
  return { error: null };
}

function parseWindSpeed(windString: string) {
  const rangeRegex = /^(?<low>\d+) to (?<high>\d+) mph$/;
  const match = windString.match(rangeRegex);
  if (match) {
    return {
      low: parseInt(match?.groups?.low || "0"),
      high: parseInt(match?.groups?.high || "0"),
    };
  } else {
    const singleRegex = /^(?<mph>\d+) mph$/;
    const match = windString.match(singleRegex);
    const range = parseInt(match?.groups?.mph || "0");
    return { low: range, high: range };
  }
}

function withinThreeDays(dateString: string): boolean {
  const startTime = Date.parse(dateString);
  const now = Date.now();
  const threeDays = 3 * 24 * 60 * 60 * 1000;
  return startTime < now + threeDays;
}

interface weatherPeriod {
  startTime: string;
  endTime: string;
  temperature: number;
  probabilityOfPrecipitation: number;
  maxWindSpeed: number;
}

function msg(g: weatherPeriod): string {
  return `${g.startTime} - ${g.endTime} is a great time to bike 🚴. Temp: ${
    g.temperature
  }, Precipitation: ${g.probabilityOfPrecipitation * 100}% Wind Speed: ${
    g.maxWindSpeed
  }`;
}

function alert(times: weatherPeriod[]) {
  const days: string[] = [];
  for (const t of times) {
    days.push(msg(t));
  }
  return `😎 Great bike weather in your near future!
  
${days.join("\n")}
  
Make a calendar entry and get out there!`;
}
function filterWeather(apiResponse: APIWeatherForecast[]): weatherPeriod[] {
  const goodTimesToBike: weatherPeriod[] = [];
  for (let period of apiResponse) {
    if (
      period.isDaytime &&
      period.temperature > 50 &&
      period.temperature < 85 &&
      period.probabilityOfPrecipitation.value < 30 &&
      parseWindSpeed(period.windSpeed).high < 15 &&
      withinThreeDays(period.startTime)
    ) {
      goodTimesToBike.push({
        startTime: period.startTime,
        endTime: period.endTime,
        temperature: period.temperature,
        probabilityOfPrecipitation: period.probabilityOfPrecipitation.value,
        maxWindSpeed: parseWindSpeed(period.windSpeed).high,
      });
    }
  }
  return goodTimesToBike;
}

async function main() {
  if (!PUSHOVER_USER || !PUSHOVER_TOKEN) {
    console.error("PUSHOVER_USER and PUSHOVER_TOKEN required");
    process.exit(1);
  }
  let { weather, error } = await getWeather();
  if (error !== "") {
    console.error(error);
    process.exit(1);
  }

  const timesToBike = filterWeather(weather);
  if (timesToBike.length > 0) {
    console.log(timesToBike);
    let { error } = await push(alert(timesToBike));
    if (error) {
      console.error(error);
      process.exit(1);
    }
  } else {
    console.log("no times found 😭");
  }
}

main();

File: weather.py
import os
import sys
import re
import json
import urllib.request


TAKOMA_PARK_FORECAST_URL = (
    "https://api.weather.gov/gridpoints/LWX/97,75/forecast/hourly"
)
PUSHOVER_USER = os.getenv("PUSHOVER_USER")
if not PUSHOVER_USER:
    print(f"PUSHOVER_USER required")
    sys.exit(1)

PUSHOVER_TOKEN = os.getenv("PUSHOVER_TOKEN")
if not PUSHOVER_TOKEN:
    print(f"PUSHOVER_TOKEN required")
    sys.exit(1)

USER_AGENT = "github.com/kingishb/good-bike-weather"
WIND_SPEED_REGEX = r"(?P<high>\d+) mph$"


# get weather forecast
with urllib.request.urlopen(TAKOMA_PARK_FORECAST_URL) as resp:
    periods = json.load(resp)["properties"]["periods"]

# find good times to bike
time_periods = []
for period in periods:
    m = re.match(WIND_SPEED_REGEX, period["windSpeed"])
    if not m:
        print("could not parse wind speed", period["windSpeed"])
        sys.exit(1)
    wind_speed = int(m["high"])
    if (
        period["isDaytime"]
        and 50 < period["temperature"] < 85
        and period["probabilityOfPrecipitation"]["value"] < 30
        and wind_speed < 15
    ):
        # merge together hourly forecast if they make up a block of good weather
        if len(time_periods) > 0 and (prev := time_periods[-1])["endTime"] == period["startTime"]:
            time_periods[-1] = {
                "startTime": prev["startTime"],
                "endTime": period["endTime"],
                "temperature": max(period["temperature"], prev["temperature"]),
                "probabilityOfPrecipitation": max(
                    period["probabilityOfPrecipitation"]["value"],
                    prev["probabilityOfPrecipitation"],
                ),
                "maxWindSpeed": max(wind_speed, prev["maxWindSpeed"]),
            }
        else:
            time_periods.append(
                {
                    "startTime": period["startTime"],
                    "endTime": period["endTime"],
                    "temperature": period["temperature"],
                    "probabilityOfPrecipitation": period["probabilityOfPrecipitation"][
                        "value"
                    ],
                    "maxWindSpeed": wind_speed,
                }
            )
if len(time_periods) == 0:
    print("😭 no times found!")
    sys.exit(0)

print(time_periods)

# build message to send
time_messages = []
for t in time_periods:
    time_messages.append(
        f'🚴 {t["startTime"]} - {t["endTime"]} Temp {t["temperature"]} F Precipitation {t["probabilityOfPrecipitation"]}% Wind Speed {t["maxWindSpeed"]} mph'
    )

t = "\n".join(time_messages)
msg = f"""☀️  Great bike weather coming up! 🚲
{t}
Make a calendar entry and get out there!"""

# send push notification
req = urllib.request.Request(
    "https://api.pushover.net/1/messages.json",
    json.dumps({"token": PUSHOVER_TOKEN, "user": PUSHOVER_USER, "message": msg}).encode(
        "utf8"
    ),
    headers={"content-type": "application/json"},
    method="POST",
)
with urllib.request.urlopen(req) as resp:
    print(json.load(resp))